### PR TITLE
Improve latency during historical sync

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes:
- When syncing historical state wait for block processing queue to be empty before handling responses.
- Move prepare request to sync version to simplify the code and reduce allocations and waits.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

**Comments about testing , should you have some** (optional)
Manual tests

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs [hunk 1]
```diff
@@ -331,6 +334,18 @@ private void FireProcessingQueueEmpty()
...
+        {
+            if (_emptiedQueueTaskCompletionSource is not null)
+            {
+                Task task = _emptiedQueueTaskCompletionSource.Task;
+                if (!task.IsCompleted)
+                {
+                    await task;
+                }
+            }
+        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:347]: Adding this method and its usages is the crucial part of this PR.

File: src/Nethermind/Nethermind.Blockchain.Test/BlockchainProcessorTests.cs [hunk 2]
```diff
@@ -365,10 +366,17 @@ public AfterBlock FullyProcessedFail(Block block)
                     .ProcessedFail(block);
             }
 
-            public ProcessingTestContext QueueIsEmpty(int count)
+            public ValueTask<ProcessingTestContext> QueueIsEmpty(int count)
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Blockchain.Test/BlockchainProcessorTests.cs:369]: Why change?

File: src/Nethermind/Nethermind.Consensus/Processing/EmptyBlockProcessingQueue.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Threading.Tasks;
+using Nethermind.Core;
+
+namespace Nethermind.Consensus.Processing;
+
+public class EmptyBlockProcessingQueue : IBlockProcessingQueue
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Consensus/Processing/EmptyBlockProcessingQueue.cs:24]: NP: Should be in a test package IMO.

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/ISyncFeed.cs [hunk 4]
```diff
@@ -26,8 +26,8 @@ public interface ISyncFeed<T>
         int FeedId { get; }
         SyncFeedState CurrentState { get; }
         event EventHandler<SyncFeedStateEventArgs> StateChanged;
-        Task<T> PrepareRequest(CancellationToken token = default);
-        SyncResponseHandlingResult HandleResponse(T response, PeerInfo peer = null);
+        T PrepareRequest(CancellationToken token = default);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Synchronization/ParallelSync/ISyncFeed.cs:29]: NP: I'd keep it to minimize change.

**Review by asdacap — APPROVED** [review]
Understandable enough. Too much change for my liking. I'd just block peer allocation, or in sync dispatcher only.

**Comment by LukaszRozmej** [comment]
> Understandable enough. Too much change for my liking. I'd just block peer allocation, or in sync dispatcher only.

Can you provide a counter-PR and we can decide based on results and usage?

**Comment by asdacap** [comment]
As it turns out, adding a dependency to SyncDispatcher was impressively hard. I ended up creating a wrapper arount sync feed, which requires changing more things. https://github.com/NethermindEth/nethermind/pull/4828

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/ISyncFeed.cs (hunk 4 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/ParallelSync/ISyncFeed.cs:16]: but it simplifies stuff.

**Comment by LukaszRozmej** [comment]
Lackluster testing results - no improvement

**Pull request closed without merging**